### PR TITLE
first attempt of mono downmix in the magnitude domain

### PR DESCRIPTION
**Pull request by faroit**
Downmixing in the magnitude domain is the recommended way for multichannel audio, since its energy preserving. Let me know if the API makes sense here.

### CODE REVIEW

File: torchaudio_contrib/functional.py [hunk 1]
```diff
@@ -47,6 +47,10 @@ def stft(signal, fft_len, hop_len, window,
     return spect
 
 
+def spectral_downmix(tensor, power=1.0):
+    return torch.sum(tensor**2/power, 1, keepdim=True) ** (power/2)
+
```
Comment (ksanjeevan) [inline torchaudio_contrib/functional.py]: Wouldn't it be `tensor**(2/power)`? I've done some timing and `torch.norm(z, p=2, dim=1, keepdim=True)` is faster for `cuda`, although then I guess the meaning of `power` changes.

Comment (keunwoochoi) [reply torchaudio_contrib/functional.py]: +1. 

I've also seen some folks talk about quite big differences of execution time between (something like) those. Not sure it'd be significant for us though.

Comment (keunwoochoi) [reply torchaudio_contrib/functional.py]: I assume `tensor.shape=(batch, channel, time and freq in the right order)` as we discussed before? The name `tensor` sounds a bit too general, I can't think of a great name though. But it'd be useful in other places as well if we coin some names for waveform batch, spectrogram batch, etc. ...Oops maybe `waveform_batch` and `spectrogram_batch`?

Comment (ksanjeevan) [reply torchaudio_contrib/functional.py]: Yeah [like](https://github.com/pytorch/pytorch/issues/5671) [these](https://github.com/pytorch/pytorch/pull/5722). Using `torch.autograd.profiler` I always find `torch.norm` faster than manual on cuda but slower on cpu...

File: torchaudio_contrib/layers.py [hunk 2]
```diff
@@ -269,7 +285,7 @@ def __repr__(self):
 
 
 def Spectrogram(fft_len=2048, hop_len=None, frame_len=None,
-                window=None, pad=0, pad_mode="reflect", power=1., **kwargs):
+                window=None, pad=0, pad_mode="reflect", power=1., mono=True, **kwargs):
```
Comment (keunwoochoi) [inline torchaudio_contrib/layers.py]: * Couldn't it be better not to automate this? 
* Also, if we follow the discussion we had at https://github.com/keunwoochoi/torchaudio-contrib/issues/31#issuecomment-476786384, maybe something like `to_mono`, `downmix_to_mono`, `downmix` rather than `mono`? We mainly talked about `resample_to`, but the same philosophy can be applied here.

**Comment by keunwoochoi** [comment]
The specific implementation seems alright in general. But probably a little more high-level question - are we gonna have waveform downmix, too? 

Ugh, out of sudden, I felt like what if we have data types of `WaveformTensor` and `SpectrumTensor` with which we can do more like functional stuff e.g.,
```
x = WaveformTensor(batch_audio)
x.spectrogram().downmix().amplitude_to_decibel()
```
Ok none of this comment is not relevant to this commit :P

File: torchaudio_contrib/layers.py (hunk 2 above)
Comment (keunwoochoi) [reply torchaudio_contrib/layers.py]: Or, just be lazy and don't change `Spectrogram` in this commit, finish the downmix itself first?

**Comment by keunwoochoi** [comment]
Hey, in #49, I tried to have this `beta_something.py`. Maybe without a test, this approach could make sense here, too? (not necessarily though)

File: torchaudio_contrib/functional.py (hunk 1 above)
Comment (f0k) [reply torchaudio_contrib/functional.py]: > I always find `torch.norm` faster than manual on cuda but slower on cpu...

We should still prefer `torch.norm` then, to benefit from possible future optimizations of that. If we opt for a particular manual formulation, then there's nothing the PyTorch developers can do to speed it up.

File: torchaudio_contrib/layers.py (hunk 2 above)
Comment (f0k) [reply torchaudio_contrib/layers.py]: I agree that this shouldn't be the default. (The naming is another issue, but in #37 you suggested we can change anything later.)

File: torchaudio_contrib/functional.py (hunk 1 above)
Comment (f0k) [reply torchaudio_contrib/functional.py]: And why should we use the power rather than the mean?

Comment (faroit) [reply torchaudio_contrib/functional.py]: > And why should we use the power rather than the mean?

mean would also work, sure.

File: torchaudio_contrib/layers.py (hunk 2 above)
Comment (faroit) [reply torchaudio_contrib/layers.py]: > I agree that this shouldn't be the default. 

this is risky since most spectrogram based tasks that do not output full spectrograms or masks but rather do classification etc would usually read stereo, downmix and perform the model on single channel. If the default downmix would be performed in the time domain using the mean, this is clearly wrong. Since in the `Spectrogram` layer we have control over the full pipeline (audio>stft>complexnorm>downmix), IMO it makes sense to put this default here instead of e.g. in the `ComplexNorm`.

File: torchaudio_contrib/functional.py [hunk 3]
```diff
@@ -47,6 +47,10 @@ def stft(signal, fft_len, hop_len, window,
     return spect
 
 
+def spectral_downmix(tensor, power=1.0):
```
Comment (hagenw) [inline torchaudio_contrib/functional.py]: In general, I would prefer to add an option `axis=1` that defaults to the expected axis in the case of `(batch, channel, time)`. But why restrict to that? it might happen that some user has `(batch, other_stuff, channel, time)`.

File: torchaudio_contrib/layers.py (hunk 2 above)
Comment (hagenw) [reply torchaudio_contrib/layers.py]: From a `Spectrogram` I would expect that it calculates the `Spectrogram`. If I put in a multi-channel signal, I would expect spectrograms for every channel. So, I would also vote for not making downmix the default behavior.

File: torchaudio_contrib/functional.py (hunk 3 above)
Comment (faroit) [reply torchaudio_contrib/functional.py]: okay, yes. I agree

File: torchaudio_contrib/layers.py (hunk 2 above)
Comment (faroit) [reply torchaudio_contrib/layers.py]: okay, I mixed up "default should not be `mono=True`" with "default should not be doing it in the Spectrogram domain".  I will set it to `False` and implement the other suggested changes soon

File: torchaudio_contrib/functional.py (hunk 3 above)
Comment (hagenw) [reply torchaudio_contrib/functional.py]: Note, this is a general comment. If we add the `axis` option here a user would expect the same for all other functionals as well. But it should be fine, if we start with `spectral_downmix` and later add it to other functionals.

File: torchaudio_contrib/layers.py (hunk 2 above)
Comment (keunwoochoi) [reply torchaudio_contrib/layers.py]: I'd even suggest removing `mono=` at all here. First of all, computing a spectrogram doesn't need to be merged with downmixing. Then there's still an issue how/where the downmix is done. Especially, if we're gonna have two downmixing methods (one on waveform, the other on spectrogram), I think we should encourage people to use them explicitly.

Comment (keunwoochoi) [reply torchaudio_contrib/layers.py]: Also, the bottom line is - let's discuss it in a separate PR.

**Comment by keunwoochoi** [comment]
Oops? Have no idea why all the commits I made to a different branch became a part of here. But thanks! Here're some comments.
- I understood we agreed that modifying `Spectrogram` API (`mono`) is a separate issue -- and indeed it is, so it should be completely removed here. 
- On the name, [I commented here](https://github.com/keunwoochoi/torchaudio-contrib/issues/54#issuecomment-498351116) and am still of the opinion (hence `SpectrumDownmix` and `WaveformDownmix`). And there, because 'waveform' is a noun, I'd prefer `Spectrum` to `Spectral`.
- For waveform downmix, given that the current torchaudio is going to be changed a lot + [this type conversion](https://github.com/pytorch/audio/blob/master/torchaudio/functional.py#L81) doesn't seem alright for me, probably it's better to implement our own functional here? Which will be simply..
```
return torch.mean(waveforms, ch_dim, True)
```
. Also, for a batch tensor, let's use a plural form!

**Comment by faroit** [comment]
> Oops? Have no idea why all the commits I made to a different branch became a part of here. But thanks! Here're some comments.

should be fine now. There was a missbehaved rebase going on after I pulled in the recent master. 

Regarding you requests, I will update those.

**Comment by faroit** [comment]
`DownmixSpectrum` would also work on `power_specgrams`. Should I rename the input for this to just `specgrams` then?

File: torchaudio_contrib/functional.py (hunk 1 above)
Comment (faroit) [reply torchaudio_contrib/functional.py]: I replaced it with mean (which also makes more sense)

**Comment by keunwoochoi** [comment]
I don't think so. By having `power=` args we're expecting the input to be `mag_specgrams`.

**Comment by keunwoochoi** [comment]
(icymi, we didn't really put it on a vote, but `DownmixSpectrum` --> `SpectrumDownmix`? especially if you agree, which is kinda voting :)

File: torchaudio_contrib/layers.py (hunk 2 above)
Comment (faroit) [reply torchaudio_contrib/layers.py]: Lets use #54 for this

**Comment by faroit** [comment]
> (icymi, we didn't really put it on a vote, but DownmixSpectrum --> SpectrumDownmix? especially if you agree, which is kinda voting :)

ha, yep, sorry... let vote first ;-)

**Comment by faroit** [comment]
are the tests added in the master stable? In that case I can provide some for the downmix functions as well

**Comment by keunwoochoi** [comment]
Yes I think so.

**Comment by keunwoochoi** [comment]
https://github.com/keunwoochoi/torchaudio-contrib/issues/54#issuecomment-499239872 We got the names :) `DownmixWaveform` and `DownmixSpectrum`.

**Comment by keunwoochoi** [comment]
@faroit Hey, I'm a bit lost, but seems like we resolved all the naming issues as well as the implementations? Is it ready to be merged now?

**Comment by faroit** [comment]
Well, I want to add unit tests but I am confused now if we could stick with pytest or not?

**Comment by keunwoochoi** [comment]
I see, +1 for unit tests. I think we have to come back to `unittest`, at least for some 'core' features that we plan to make a PR there, and this one would be definitely core.

File: torchaudio_contrib/functional.py [hunk 4]
```diff
@@ -103,6 +103,30 @@ def stft(waveforms, fft_len, hop_len, window,
     return complex_specgrams
 
 
+def waveform_downmix(waveforms, ch_dim=1):
```
Comment (f0k) [inline torchaudio_contrib/functional.py]: I'd just call it `dim`. There is only one dimension to specify, and this is consistent with pytorch. Maybe someone may want to downmix across the batch dimension. Make sure it's included in the docstring, though. The docstring should also mention that it's downmixing by taking the mean.

File: torchaudio_contrib/layers.py [hunk 5]
```diff
@@ -118,6 +119,36 @@ def __repr__(self):
         return self.__class__.__name__ + '(power={})'.format(self.power)
 
 
+class WaveformDownmix(nn.Module):
+    """
+    Wrap torchaudio_contrib.waveform_downmix in an nn.Module.
+    """
+
+    def __init__(self):
```
Comment (f0k) [inline torchaudio_contrib/layers.py]: The `dim` option should be available (and documented) here as well.
